### ISSUE TITLE
libcurl 8.9.1 causes abrupt exit due to SIGPIPE

### ISSUE THREAD

**Issue by imaami** [transmission/transmission#7035]
### What is the issue?

transmission-daemon 4.0.6 (38c164933e) began abruptly exiting with process exit code 141 (128 + `SIGPIPE`) after upgrading libcurl to 8.9.1. There is an apparent regression in cURL causing `SIGPIPE` to not be blocked during connection shutdown handling to the same extent as before (see https://github.com/curl/curl/issues/14344).

It's probably fair to say that this is on libcurl, not transmission, but there might be benefit in transmission also being proactively mindful of errant signals.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.6 (38c164933e)

**Comment by maztheman** [comment]
This prevents transmission from working on my container.

**Comment by MoonSweep** [comment]
It also affects `transmission-gtk`.

@imaami Thanks a lot for your report ! I was struggling with this problem for a couple of days, playing with `gdb` and getting nowhere, but downgrading `curl` packages to the previous version (8.9.0-3) allowed me to get Transmission working again. I'll report a bug in Debian BTS.

**Comment by habys** [comment]
Looks like [this](https://github.com/curl/curl/issues/14344) is this issue.

**Comment by darkshram** [comment]
Please, try building **curl 8.9.1** with the following patch. It worked for me. Have been using transmission and qbittorrent (also affected) without any issues since I applied it to curl. It reverts https://github.com/curl/curl/pull/14296, which happens to be the origin of the issue.

```
diff --git a/lib/conncache.c b/lib/conncache.c
index 5de10de91..e4396b0bc 100644
--- a/lib/conncache.c
+++ b/lib/conncache.c
@@ -584,15 +584,15 @@ static void connc_close_all(struct conncache *connc)
     return;
 
   /* Move all connections to the shutdown list */
-  sigpipe_init(&pipe_st);
   conn = connc_find_first_connection(connc);
   while(conn) {
     connc_remove_conn(connc, conn);
-    sigpipe_apply(data, &pipe_st);
+    sigpipe_ignore(data, &pipe_st);
     /* This will remove the connection from the cache */
     connclose(conn, "kill all");
     Curl_conncache_remove_conn(connc->closure_handle, conn, TRUE);
     connc_discard_conn(connc, connc->closure_handle, conn, FALSE);
+    sigpipe_restore(&pipe_st);
 
     conn = connc_find_first_connection(connc);
   }
@@ -613,7 +613,7 @@ static void connc_close_all(struct conncache *connc)
   /* discard all connections in the shutdown list */
   connc_shutdown_discard_all(connc);
 
-  sigpipe_apply(data, &pipe_st);
+  sigpipe_ignore(data, &pipe_st);
   Curl_hostcache_clean(data, data->dns.hostcache);
   Curl_close(&data);
   sigpipe_restore(&pipe_st);
@@ -628,6 +628,7 @@ static void connc_shutdown_discard_oldest(struct conncache *connc)
 {
   struct Curl_llist_element *e;
   struct connectdata *conn;
+  SIGPIPE_VARIABLE(pipe_st);
 
   DEBUGASSERT(!connc->shutdowns.iter_locked);
   if(connc->shutdowns.iter_locked)
@@ -635,11 +636,9 @@ static void connc_shutdown_discard_oldest(struct conncache *connc)
 
   e = connc->shutdowns.conn_list.head;
   if(e) {
-    SIGPIPE_VARIABLE(pipe_st);
     conn = e->ptr;
     Curl_llist_remove(&connc->shutdowns.conn_list, e, NULL);
-    sigpipe_init(&pipe_st);
-    sigpipe_apply(connc->closure_handle, &pipe_st);
+    sigpipe_ignore(connc->closure_handle, &pipe_st);
     connc_disconnect(NULL, conn, connc, FALSE);
     sigpipe_restore(&pipe_st);
   }
@@ -899,9 +898,6 @@ static void connc_perform(struct conncache *connc)
   struct Curl_llist_element *e = connc->shutdowns.conn_list.head;
   struct Curl_llist_element *enext;
   struct connectdata *conn;
-  struct curltime *nowp = NULL;
-  struct curltime now;
-  timediff_t next_from_now_ms = 0, ms;
   bool done;
 
   if(!e)
@@ -924,22 +920,9 @@ static void connc_perform(struct conncache *connc)
       Curl_llist_remove(&connc->shutdowns.conn_list, e, NULL);
       connc_disconnect(NULL, conn, connc, FALSE);
     }
-    else {
-      /* Not done, when does this connection time out? */
-      if(!nowp) {
-        now = Curl_now();
-        nowp = &now;
-      }
-      ms = Curl_conn_shutdown_timeleft(conn, nowp);
-      if(ms && ms < next_from_now_ms)
-        next_from_now_ms = ms;
-    }
     e = enext;
   }
   connc->shutdowns.iter_locked = FALSE;
-
-  if(next_from_now_ms)
-    Curl_expire(data, next_from_now_ms, EXPIRE_RUN_NOW);
 }
 
 void Curl_conncache_multi_perform(struct Curl_multi *multi)
diff --git a/lib/connect.c b/lib/connect.c
index 9b68f15da..604162926 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -161,7 +161,6 @@ timediff_t Curl_shutdown_timeleft(struct connectdata *conn, int sockindex,
                                   struct curltime *nowp)
 {
   struct curltime now;
-  timediff_t left_ms;
 
   if(!conn->shutdown.start[sockindex].tv_sec || !conn->shutdown.timeout_ms)
     return 0; /* not started or no limits */
@@ -170,30 +169,8 @@ timediff_t Curl_shutdown_timeleft(struct connectdata *conn, int sockindex,
     now = Curl_now();
     nowp = &now;
   }
-  left_ms = conn->shutdown.timeout_ms -
-            Curl_timediff(*nowp, conn->shutdown.start[sockindex]);
-  return left_ms? left_ms : -1;
-}
-
-timediff_t Curl_conn_shutdown_timeleft(struct connectdata *conn,
-                                       struct curltime *nowp)
-{
-  timediff_t left_ms = 0, ms;
-  struct curltime now;
-  int i;
-
-  for(i = 0; conn->shutdown.timeout_ms && (i < 2); ++i) {
-    if(!conn->shutdown.start[i].tv_sec)
-      continue;
-    if(!nowp) {
-      now = Curl_now();
-      nowp = &now;
-    }
-    ms = Curl_shutdown_timeleft(conn, i, nowp);
-    if(ms && (!left_ms || ms < left_ms))
-      left_ms = ms;
-  }
-  return left_ms;
+  return conn->shutdown.timeout_ms -
+         Curl_timediff(*nowp, conn->shutdown.start[sockindex]);
 }
 
 void Curl_shutdown_clear(struct Curl_easy *data, int sockindex)
diff --git a/lib/connect.h b/lib/connect.h
index e9a4f9c59..ffceeb87f 100644
--- a/lib/connect.h
+++ b/lib/connect.h
@@ -46,15 +46,10 @@ void Curl_shutdown_start(struct Curl_easy *data, int sockindex,
                          struct curltime *nowp);
 
 /* return how much time there is left to shutdown the connection at
- * sockindex. Returns 0 if there is no limit or shutdown has not started. */
+ * sockindex. */
 timediff_t Curl_shutdown_timeleft(struct connectdata *conn, int sockindex,
                                   struct curltime *nowp);
 
-/* return how much time there is left to shutdown the connection.
- * Returns 0 if there is no limit or shutdown has not started. */
-timediff_t Curl_conn_shutdown_timeleft(struct connectdata *conn,
-                                       struct curltime *nowp);
-
 void Curl_shutdown_clear(struct Curl_easy *data, int sockindex);
 
 /* TRUE iff shutdown has been started */
diff --git a/lib/easy.c b/lib/easy.c
index 296dae8c1..ed4bc357e 100644
--- a/lib/easy.c
+++ b/lib/easy.c
@@ -764,8 +764,7 @@ static CURLcode easy_perform(struct Curl_easy *data, bool events)
   /* assign this after curl_multi_add_handle() */
   data->multi_easy = multi;
 
-  sigpipe_init(&pipe_st);
-  sigpipe_apply(data, &pipe_st);
+  sigpipe_ignore(data, &pipe_st);
 
   /* run the transfer */
   result = events ? easy_events(multi) : easy_transfer(multi);
diff --git a/lib/multi.c b/lib/multi.c
index 1461bc441..6b8bef3aa 100644
--- a/lib/multi.c
+++ b/lib/multi.c
@@ -2706,7 +2706,6 @@ CURLMcode curl_multi_perform(struct Curl_multi *multi, int *running_handles)
   CURLMcode returncode = CURLM_OK;
   struct Curl_tree *t;
   struct curltime now = Curl_now();
-  SIGPIPE_VARIABLE(pipe_st);
 
   if(!GOOD_MULTI_HANDLE(multi))
     return CURLM_BAD_HANDLE;
@@ -2714,10 +2713,12 @@ CURLMcode curl_multi_perform(struct Curl_multi *multi, int *running_handles)
   if(multi->in_callback)
     return CURLM_RECURSIVE_API_CALL;
 
-  sigpipe_init(&pipe_st);
   data = multi->easyp;
   if(data) {
     CURLMcode result;
+    bool nosig = data->set.no_signal;
+    SIGPIPE_VARIABLE(pipe_st);
+    sigpipe_ignore(data, &pipe_st);
     /* Do the loop and only alter the signal ignore state if the next handle
        has a different NO_SIGNAL state than the previous */
     do {
@@ -2725,23 +2726,22 @@ CURLMcode curl_multi_perform(struct Curl_multi *multi, int *running_handles)
          pointer now */
       struct Curl_easy *datanext = data->next;
 
-    if(data != multi->conn_cache.closure_handle) {
-        /* connection cache handle is processed below */
-        sigpipe_apply(data, &pipe_st);
-        result = multi_runsingle(multi, &now, data);
-        if(result)
-          returncode = result;
+      if(data->set.no_signal != nosig) {
+        sigpipe_restore(&pipe_st);
+        sigpipe_ignore(data, &pipe_st);
+        nosig = data->set.no_signal;
       }
+      result = multi_runsingle(multi, &now, data);
+      if(result)
+        returncode = result;
 
       data = datanext; /* operate on next handle */
     } while(data);
+    sigpipe_restore(&pipe_st);
   }
 
-  sigpipe_apply(multi->conn_cache.closure_handle, &pipe_st);
   Curl_conncache_multi_perform(multi);
 
-  sigpipe_restore(&pipe_st);
-
   /*
    * Simply remove all expired timers from the splay since handles are dealt
    * with unconditionally by this function and curl_multi_timeout() requires
@@ -3181,7 +3181,8 @@ static CURLMcode multi_socket(struct Curl_multi *multi,
   struct Curl_easy *data = NULL;
   struct Curl_tree *t;
   struct curltime now = Curl_now();
-  bool run_conn_cache = FALSE;
+  bool first = FALSE;
+  bool nosig = FALSE;
   SIGPIPE_VARIABLE(pipe_st);
 
   if(checkall) {
@@ -3226,15 +3227,11 @@ static CURLMcode multi_socket(struct Curl_multi *multi,
         DEBUGASSERT(data);
         DEBUGASSERT(data->magic == CURLEASY_MAGIC_NUMBER);
 
-        if(data == multi->conn_cache.closure_handle)
-          run_conn_cache = TRUE;
-        else {
-          if(data->conn && !(data->conn->handler->flags & PROTOPT_DIRLOCK))
-            /* set socket event bitmask if they are not locked */
-            data->state.select_bits |= (unsigned char)ev_bitmask;
+        if(data->conn && !(data->conn->handler->flags & PROTOPT_DIRLOCK))
+          /* set socket event bitmask if they are not locked */
+          data->state.select_bits |= (unsigned char)ev_bitmask;
 
-          Curl_expire(data, 0, EXPIRE_RUN_NOW);
-        }
+        Curl_expire(data, 0, EXPIRE_RUN_NOW);
       }
 
       /* Now we fall-through and do the timer-based stuff, since we do not want
@@ -3261,13 +3258,19 @@ static CURLMcode multi_socket(struct Curl_multi *multi,
    * to process in the splay and 'data' will be re-assigned for every expired
    * handle we deal with.
    */
-  sigpipe_init(&pipe_st);
   do {
-    if(data == multi->conn_cache.closure_handle)
-      run_conn_cache = TRUE;
     /* the first loop lap 'data' can be NULL */
-    else if(data) {
-      sigpipe_apply(data, &pipe_st);
+    if(data) {
+      if(!first) {
+        first = TRUE;
+        nosig = data->set.no_signal; /* initial state */
+        sigpipe_ignore(data, &pipe_st);
+      }
+      else if(data->set.no_signal != nosig) {
+        sigpipe_restore(&pipe_st);
+        sigpipe_ignore(data, &pipe_st);
+        nosig = data->set.no_signal; /* remember new state */
+      }
       result = multi_runsingle(multi, &now, data);
 
       if(CURLM_OK >= result) {
@@ -3289,13 +3292,8 @@ static CURLMcode multi_socket(struct Curl_multi *multi,
     }
 
   } while(t);
-
-  if(run_conn_cache) {
-    sigpipe_apply(multi->conn_cache.closure_handle, &pipe_st);
-    Curl_conncache_multi_perform(multi);
-  }
-
-  sigpipe_restore(&pipe_st);
+  if(first)
+    sigpipe_restore(&pipe_st);
 
   if(running_handles)
     *running_handles = (int)multi->num_alive;
diff --git a/lib/sigpipe.h b/lib/sigpipe.h
index b91a2f513..9b29403c2 100644
--- a/lib/sigpipe.h
+++ b/lib/sigpipe.h
@@ -36,11 +36,6 @@ struct sigpipe_ignore {
 
 #define SIGPIPE_VARIABLE(x) struct sigpipe_ignore x
 
-static void sigpipe_init(struct sigpipe_ignore *ig)
-{
-  memset(ig, 0, sizeof(*ig));
-}
-
 /*
  * sigpipe_ignore() makes sure we ignore SIGPIPE while running libcurl
  * internals, and then sigpipe_restore() will restore the situation when we
@@ -75,20 +70,9 @@ static void sigpipe_restore(struct sigpipe_ignore *ig)
     sigaction(SIGPIPE, &ig->old_pipe_act, NULL);
 }
 
-static void sigpipe_apply(struct Curl_easy *data,
-                          struct sigpipe_ignore *ig)
-{
-  if(data->set.no_signal != ig->no_signal) {
-    sigpipe_restore(ig);
-    sigpipe_ignore(data, ig);
-  }
-}
-
 #else
 /* for systems without sigaction */
 #define sigpipe_ignore(x,y) Curl_nop_stmt
-#define sigpipe_apply(x,y) Curl_nop_stmt
-#define sigpipe_init(x)  Curl_nop_stmt
 #define sigpipe_restore(x)  Curl_nop_stmt
 #define SIGPIPE_VARIABLE(x)
 #endif
```

**Comment by BGazotti** [comment]
Looks like a curl issue. Rebuilt 8.9.1 with https://gitweb.gentoo.org/repo/gentoo.git/tree/net-misc/curl/files/curl-8.9.1-sigpipe.patch and it seems to have fixed it as well.
```
--- a/lib/sigpipe.h
+++ b/lib/sigpipe.h
@@ -39,6 +39,7 @@ struct sigpipe_ignore {
 static void sigpipe_init(struct sigpipe_ignore *ig)
 {
   memset(ig, 0, sizeof(*ig));
+  ig->no_signal = TRUE;
 }
 
 /*
```

**Comment by imaami** [comment]
> Looks like a curl issue. Rebuilt 8.9.1 with https://gitweb.gentoo.org/repo/gentoo.git/tree/net-misc/curl/files/curl-8.9.1-sigpipe.patch and it seems to have fixed it as well.
> 
> ```
> --- a/lib/sigpipe.h
> +++ b/lib/sigpipe.h
> @@ -39,6 +39,7 @@ struct sigpipe_ignore {
>  static void sigpipe_init(struct sigpipe_ignore *ig)
>  {
>    memset(ig, 0, sizeof(*ig));
> +  ig->no_signal = TRUE;
>  }
>  
>  /*
> ```

This is indeed the upstream fix (https://github.com/curl/curl/commit/3eec5afbd0b6377eca893c392569b2faf094d970) and a much better choice compared to just reverting the original commit. 8.9.0 had a problem with high CPU utilization, and the `SIGPIPE` bug was an unfortunate regression from fixing that. 8.9.2 will be released very soon, I assume.

**Comment by Czechball** [comment]
> > Looks like a curl issue. Rebuilt 8.9.1 with https://gitweb.gentoo.org/repo/gentoo.git/tree/net-misc/curl/files/curl-8.9.1-sigpipe.patch and it seems to have fixed it as well.
> > ```
> > --- a/lib/sigpipe.h
> > +++ b/lib/sigpipe.h
> > @@ -39,6 +39,7 @@ struct sigpipe_ignore {
> >  static void sigpipe_init(struct sigpipe_ignore *ig)
> >  {
> >    memset(ig, 0, sizeof(*ig));
> > +  ig->no_signal = TRUE;
> >  }
> >  
> >  /*
> > ```
> 
> This is indeed the upstream fix ([curl/curl@3eec5af](https://github.com/curl/curl/commit/3eec5afbd0b6377eca893c392569b2faf094d970)) and a much better choice compared to just reverting the original commit. 8.9.0 had a problem with high CPU utilization, and the `SIGPIPE` bug was an unfortunate regression from fixing that. ~8.9.2 will be released very soon, I assume.~ <- scratch the last one, it'll be 8.10.0

Doesn't seem to work for me. Built curl 8.10.0-DEV from https://github.com/curl/curl/commit/82bbb386a30f58aae071dd8e57076e0557a3b474, installed, curl works fine. Transmission keeps exiting with the same behaviour as with curl 8.9.1 from upstream

**Comment by Czechball** [comment]
Update: Seems like it was my mistake, I only updated curl but not libcurl, which stayed at 8.9.1. With both curl and libcurl updated to https://github.com/curl/curl/commit/82bbb386a30f58aae071dd8e57076e0557a3b474, transmission seems stable.

**Comment by ivanjx** [comment]
latest 4.0.6 is completely unstable rendering it unusable. reverting to older version now.
any estimate on when the fix would be released? thanks

**Comment by tavicu** [comment]
Trackers are starting to remove Transmission v4.0 from whitelists. Any news on this?

**Comment by BowlesCR** [comment]
+1 for Transmission being more defensive about this where possible, but the real issue here is with the distro/packaging being too far ahead of stable.

From a quick survey, I've only found bleeding edge distros like Arch, Gentoo, and Alpine:edge to be impacted, and Gentoo appears to have a patch out already. Fedora:rawhide is notably **not** impacted.
linuxserver/transmission is built on alpine:edge -- which is a questionable decision.

**Comment by radupotop** [comment]
Arch has published a patched version of curl as well.

**Comment by Roardom** [comment]
> Trackers are starting to remove Transmission v4.0 from whitelists. Any news on this?

The reason it is currently temporarily being banned is because the continual restarting of the client session (caused by I'm assuming docker containers or other daemons) creates a new session on the tracker every time, creating many duplicate peers saved in the database (many trackers uniquely identify peers by the combination of the peer_id query parameter, user id and torrent id). This creates a new database record every announce instead of updating existing entries, as well as prevents calculating the user upload and download deltas from the previous announce since it uses a different unique key. For trackers to whitelist transmission again, the client session should not be continuously restarted, or should send a `stopped` event before every restart.

**Comment by dechamps** [comment]
> the continual restarting of the client session (caused by I'm assuming docker containers or other daemons)

This would seem to suggest that auto-restarting a BitTorrent client is bad practice and should be removed from standard configs. Transmission is not supposed to crash; if it does then hiding the problem with automatic restarts is unlikely to be the best course of action. Crash loops like can lead to accidental DDoS of trackers.

**Comment by Roardom** [comment]
> > the continual restarting of the client session (caused by I'm assuming docker containers or other daemons)
> 
> This would seem to suggest that auto-restarting a BitTorrent client is bad practice and should be removed from standard configs. Transmission is not supposed to crash; if it does then hiding the problem with automatic restarts is unlikely to be the best course of action. Crash loops like can lead to accidental DDoS of trackers.

That is correct. A fairly large tracker was suffering many 502's yesterday which they tracked down to this bug before blacklisting this version of transmission so that they can remain online.

What's more important than continual starting however is that the client informs the tracker that they've ended their previous client session, so that it can be removed from memory and the stats between the last announce and the shut down are kept track of (obviously I understand this is impossible in the event of a crash). Otherwise, trackers are generally coded to keep these sessions around as long as possible to ensure more accurate statistics (Trade off between accuracy and performance).

**Comment by reardonia** [comment]
@tearfur @ckerr : should we consider making peer-id something slightly more stable? 

This issue has lead to 4.0.6 getting banned at several private trackers, because apparently in several distributions it is set to auto-restart on crash.  This in turns leads to generating new peer-id and effectively flooding tracker with announces.  The announce flood would be bad on its own (and would not be unique to Transmission--any client crashing and restarting will cause this) but what makes it worse is that a new peer-id is generated on each reload of the app/daemon. From tracker respective, it looks like a new client each time.

Perhaps the peer-id could be retained in resume file, and only regenerated after 24h?

**Comment by reardonia** [comment]
> > Trackers are starting to remove Transmission v4.0 from whitelists. Any news on this?
> 
> The reason it is currently temporarily being banned is because the continual restarting of the client session (caused by I'm assuming docker containers or other daemons) creates a new session on the tracker every time, creating many duplicate peers saved in the database (many trackers uniquely identify peers by the combination of the peer_id query parameter, user id and torrent id). This creates a new database record every announce instead of updating existing entries, as well as prevents calculating the user upload and download deltas from the previous announce since it uses a different unique key. For trackers to whitelist transmission again, the client session should not be continuously restarted, or should send a `stopped` event before every restart.

The probem with sending a "stopped" event is that the prior peer-id has already been lost.

**Comment by reardonia** [comment]
> > the continual restarting of the client session (caused by I'm assuming docker containers or other daemons)
> 
> This would seem to suggest that auto-restarting a BitTorrent client is bad practice and should be removed from standard configs. Transmission is not supposed to crash; if it does then hiding the problem with automatic restarts is unlikely to be the best course of action. Crash loops like these can lead to accidental DDoS of trackers.

The .service file here does not autorestart, so this must be specific to a proprietary distribution?  Does anyone have the service configuration file for the environment that was causing the problem?

**Comment by habys** [comment]
It definitely does not restart in a troublesome way by default on Arch.

**Comment by imaami** [comment]
> This issue has lead to 4.0.6 getting banned at several private trackers

I don't know if these bans are based on the client's self-identification and the version number in it, but if so it's effectively a false positive. Most instances out there are dynamically linked, and using a patched libcurl (or the next libcurl version *once it's released*) fixes the problem. Transmission version number 4.0.6 doesn't indicate the presence of the bug, it just happens to correlate fairly strongly with a buggy libcurl *right now at this specific moment in time*.

Transmission 4.0.6 stops "being buggy" when the system libcurl package is updated to a patched one or downgraded to 8.9.0.

**Comment by imaami** [comment]
> It definitely does not restart in a troublesome way by default on Arch.

Assuming your libcurl package is version 8.9.1-2 then the reason it works is because of this: ttps://gitlab.archlinux.org/archlinux/packaging/packages/curl/-/commit/78ea3389296d19944fdbb7869e46253c87301d11

**Comment by nastasie-octavian** [comment]
Arch Linux uses the upstream transmission-daemon.service and that doesn't have a Restart option. Also, when I upgraded to libcurl 8.9.1 and Transmission started crashing on Arch Linux, I had to manually restart the service after every crash to restore the web client, so I'm pretty sure that, _independent of the curl bug_, Arch Linux does not automatically restart Transmission when it crashes.

As for private trackers banning Transmission, some of them are **banning 4.X in general**, not just 4.0.6, and I'm not sure how much they're going to check/care about the specifics of _why_ Transmission is causing them issues, i.e. because of a bug in curl 8.9.1 in this instance. They're probably (?) going to wait until a new 4.0.X Transmission version is out which is guaranteed (?) not to cause them issues to (re)whitelist Transmission again, at least from that new version onwards.

A more robust handling of the client session when Transmission crashes either because of a bug in a library it depends on or because of a bug in Transmission itself would be a better and more future proof solution to prevent Transmission being added to blacklists. Granted, that's a feature outside the scope of this particular curl issue.

**Comment by tearfur** [comment]
> It's probably fair to say that this is on libcurl, not transmission, but there might be benefit in transmission also being proactively mindful of errant signals.

I think we are already doing that. Or did you capture a stack trace showing SIGPIPE being triggered by Transmission's code?

https://github.com/transmission/transmission/blob/38c164933e9f77c110b48fe745861c3b98e3d83e/libtransmission/session-thread.cc#L225-L226

https://github.com/transmission/transmission/blob/38c164933e9f77c110b48fe745861c3b98e3d83e/libtransmission/session.cc#L638-L639

**Comment by tearfur** [comment]
Hmm, I cannot figure out why this happened even if we ignored `SIGPIPE`...

> Perhaps the peer-id could be retained in resume file, and only regenerated after 24h?

@reardonia This doesn't sound too bad to me.

Just to be safe, anyone can think of drawbacks of this?

**Comment by zwimer** [comment]
Given that 4.0.* is seemingly being banned on multiple trackers, is there any knowledge when 4.1.0 might be released? Or is that still a long way away? I do see the milestone is at 89%.

**Comment by reardonia** [comment]
> Given that 4.0.* is seemingly being banned on multiple trackers, is there any knowledge when 4.1.0 might be released? Or is that still a long way away? I do see the milestone is at 89%.

Very much hoping that the maintainers are able to find time soon to address the significant backlog of PR's. A lot of time&energy has been spent addressing major problems that crept in during the C++ rewrite, but those fixes are stuck in review and we don't have any indication from maintainers as to when they will start working through the backlog.

**Comment by dechamps** [comment]
Another Transmission code change that may be useful is changing the Transmission user agent:

https://github.com/transmission/transmission/blob/3e9f5f614a875aae987235d3836d0ca555a80d80/libtransmission/session.cc#L311-L314

To also include the libcurl version. This would allow trackers to do more targeted banning of Transmission clients that use a known buggy libcurl, as opposed to the blanket overbroad banning we're seeing now.

**Comment by ivanjx** [comment]
> To also include the libcurl version.

what if the future crashes are caused by different libraries?

**Comment by foobarna** [comment]
Just for my understanding: a build [releases page](https://github.com/transmission/transmission/releases), such as  [Transmission-4.0.6.dmg]https://github.com/transmission/transmission/releases/download/4.0.6/Transmission-4.0.6.dmg) should not be affected, right?

**Comment by dechamps** [comment]
> what if the future crashes are caused by different libraries?

I understand your point, but I think libcurl version is especially important to have there because libcurl is an integral part of these HTTP requests that the User-Agent is being reported in, and is also arguably the most important/complex/failure-prone dependency out of all the libraries Transmission directly depends on. If we were to have one library version in there it would clearly be that one.

Also, this is not just about the future - if done quickly enough it could help mitigating the current "accidental DDoS attack", as trackers could start allowing any Transmission client that is explicitly advertising any libcurl that is not 8.9.1.

**Comment by ivanjx** [comment]
@foobarna im not sure how dmg works but if it does include libcurl then i dont think it has the bug. if it doesnt include the libcurl then just dont update to latest.

**Comment by reardonia** [comment]
> > To also include the libcurl version.
> 
> what if the future crashes are caused by different libraries?

Not all libraries are created equal...in order, I would argue that the sensitivities should be with libevent, libcurl, and libssl/libcrypto (or whatever crypto is used).  Browser user-agents have included a comically long list of qualifiers since the 90's.

I would vote for including those three libraries in the user-agent.

**Comment by reardonia** [comment]
> > It's probably fair to say that this is on libcurl, not transmission, but there might be benefit in transmission also being proactively mindful of errant signals.
> 
> @imaami I think we are already doing that. Or did you capture a stack trace showing SIGPIPE being triggered by Transmission's code?
> 
> https://github.com/transmission/transmission/blob/38c164933e9f77c110b48fe745861c3b98e3d83e/libtransmission/session-thread.cc#L225-L226
> 
> https://github.com/transmission/transmission/blob/38c164933e9f77c110b48fe745861c3b98e3d83e/libtransmission/session.cc#L638-L639

Don't actually understand signal handling well enough... is it per-thread?  Does the web-thread inherit the context to ignore SIGPIPE?

**Comment by imaami** [comment]
> > It's probably fair to say that this is on libcurl, not transmission, but there might be benefit in transmission also being proactively mindful of errant signals.
> 
> @imaami I think we are already doing that. Or did you capture a stack trace showing SIGPIPE being triggered by Transmission's code?
> 
> https://github.com/transmission/transmission/blob/38c164933e9f77c110b48fe745861c3b98e3d83e/libtransmission/session-thread.cc#L225-L226
> 
> https://github.com/transmission/transmission/blob/38c164933e9f77c110b48fe745861c3b98e3d83e/libtransmission/session.cc#L638-L639

`signal()` is incompatible with multi-threaded code (its behavior is unspecified), and `sigaction()` (which glibc uses to implement `signal()`) only sets the handler for the calling thread. At least on Linux the proper way to guarantee no thread is able to receive signals unless explicitly allowed is to call `pthread_sigmask()` from the main thread before starting any threads at all. That way every thread you create inherits the mask you want. Then you selectively unblock what you want inside threads that you know should receive them.

**Comment by namazso** [comment]
Including the version of everything (transmission, all libs, OS, maybe kernel version?) in the user agent sounds like could fix the trackers incorrectly banning random versions for irrelevant issues.

Though this is assuming they care, from comments here it seems the trackers didn't ban 3.x even though it looks like it has the same issue as long as it's using the buggy libcurl version.

**Comment by garoto** [comment]
> [...] seems the trackers didn't ban 3.x even though it looks like it has the same issue as long as it's using the buggy libcurl version.

Pretty much yeah. My transmission v4.0.6 was compiled against a very old libcurl-dev (v7.64.0, from Buster repos), still got banned in two trackers I'm in. One of those had all 4.x.x versions banned, but that has since being remedied and only v4.0.6 is currently blacklisted.

Anyways, I imagine that ideally a hash of the client source plus some of the important libs the program is compiled against could be made part of the peerid string.

Would probably be too convoluted and error prone, but a way to differentiate between compiles would come in handy it seems.

**Comment by dechamps** [comment]
> Though this is assuming they care, from comments here it seems the trackers didn't ban 3.x even though it looks like it has the same issue as long as it's using the buggy libcurl version.

Tracker operators are well aware this is a libcurl bug, not a transmission bug. But they have no way of targetting the libcurl version, so they are banning Transmission 4.0.6 with the rationale that the latest version of Transmission is highly correlated with the highest version of libcurl. Obviously this implies there will be some false positives and some false negatives, but they are doing what they can to stop the DDoS from auto-restarting broken libcurl clients.

**Comment by garoto** [comment]
Excellent `tl;dr` @dechamps. Nailed.

**Comment by Roardom** [comment]
> Hmm, I cannot figure out why this happened even if we ignored `SIGPIPE`...
> 
> > Perhaps the peer-id could be retained in resume file, and only regenerated after 24h?
> 
> @reardonia This doesn't sound too bad to me.
> 
> Just to be safe, anyone can think of drawbacks of this?

A few things I thought of:

Trackers include the peer_id in the primary key so that users can seed the same torrent from multiple instances (most will also limit the number of instances at a time to ~3 or so). It has to be ensured that no instance ever shares the same peer id on a torrent with another instance (or be highly improbable).

The peer_id is directly tied to stats provided by the client. As mentioned previously, most trackers calculate a user's upload/download by finding the previous announce using the same user_id, torrent_id, peer_id, and finding the difference of upload/download between them, and increasing the user's stats by the delta. Because of this, if the same peer id is used for new sessions where the `uploaded`/`downloaded` stats are reset back to 0, then it will mess up the stats on the tracker when calculating the delta (although many trackers account for this and assume the delta is `0` if it calculates a negative number). If stats are being reset to zero, then the peer_id should be changed.

Some trackers also make the assumption that if a peer hasn't announced since their `interval` has passed, that the peer is gone and can be deleted. However, other trackers assume the possibility of a user's internet connection being flaky in hopes of the peer eventually coming back online and reporting their stats and will wait a certain amount of time before permanently deleting the peer, but marking the peer as inactive to not send to peer lists. Many trackers also assume that when a peer is stored for the first time, that the upload/download stats are 0, to reduce the impact of a peer returning with non-zero upload/download after the peer was deleted. Given that there's nothing in the BitTorrent spec that says how long trackers should wait before deleting a peer, I'm not too sure how re-using the same peer_id with its associated stats after a long pause in announcing should be handled.

**Comment by BowlesCR** [comment]
> The .service file here does not autorestart, so this must be specific to a proprietary distribution? Does anyone have the service configuration file for the environment that was causing the problem?

The largest culprit seems to have been https://github.com/linuxserver/docker-transmission -- which has `restart: unless-stopped` in the config, and is based on alpine:edge.

**Comment by dechamps** [comment]
> The largest culprit seems to have been https://github.com/linuxserver/docker-transmission -- which has `restart: unless-stopped` in the config, and is based on alpine:edge.

Yeah, I agree it is quite plausible: many people running auto-updating, auto-restarting Transmission Dockers on NAS devices and stuff, which then end up forming an accidental botnet. I've already filed linuxserver/docker-transmission#287 about that. The response has been… underwhelming, to say the least. If someone wants to reopen a separate issue to try to get them to understand that ignoring reports of DDoS attacks may not be in anyone's best interests, be my guest.

**Comment by ivanjx** [comment]
are the bans only from private trackers? most of my trackers does not seem to ban me until now. i use the lsio docker and it has been fixed since yesterday.

**Comment by tearfur** [comment]
> `signal()` is incompatible with multi-threaded code (its behavior is unspecified)

@imaami AFAIK signals are supposed to be process-wide. From `pthreads(7)`:

> POSIX.1 also requires that threads share a range of other attributes (i.e., these attributes are process-wide rather than per-thread):
>
> ...
>
> - signal dispositions

Here is a test program to for this. On my machine, it returns normally, which matches what the `pthreads` man page says.

<details>
<summary>test code</summary>

```cpp
#include <array>
#include <csignal>
#include <thread>

#include <unistd.h>

void sigpipe_ignore()
{
    (void)signal(SIGPIPE, SIG_IGN);
}

int main() {
    auto s = std::array<int, 2>{};
    pipe(std::data(s));
    close(s[0]);
    auto t = std::thread(sigpipe_ignore);
    t.join();
    write(s[1], std::data(s), std::size(s) * sizeof(int));
    close(s[1]);
    return 0;
}
```

</details>

> If stats are being reset to zero, then the peer_id should be changed.

@Roardom Hmm, then sounds like keeping the peer-id code as-is is the best option for now.

**Comment by vincejv** [comment]
For anyone looking to migrate to QBittorrent or clients based on arvidn's libtorrent please see possible issues during migrations due to libtorrent stripping of some non-printable characters (for good reason, security wise), you'll be encountering issues like File is Missing/IO Errors

transmission leaves this security hole unpatched, just migrate to qbt or rtorrent already

Context:
- https://github.com/arvidn/libtorrent/issues/7568

**Comment by heit0r** [comment]
Downgrading curl to version 8.7.1 (not 8.9.0) has got transmission stable here, with no more abrupt exits. I don't feel like Transmission has been blacklisted in the tracker I usually seed. Everything came back to normal, although this is just a temporary solution. Thank you @imaami for bringing attention on this issue.

**Comment by heit0r** [comment]
Downgrading curl to version 8.7.1 (not 8.9.0) has got transmission stable here, with no more abrupt exits. Transmission was exiting with less than a minute after it starts. Now I just have it opened for half an hour and seeding normally. I don't feel like Transmission has been blacklisted at least in the tracker I usually seed. Everything came back to normal, although this is just a temporary solution. I'm just waiting for the new release of libcurl here. Thank you @imaami for bringing attention on this issue.

**Comment by Chaotic** [comment]
Transmission 4.0.6 has been blacklisted on at least one of the trackers I use regularly. After it was already fixed as well.  I don't suppose we could get a 4.0.7 release to get past that version issue.

**Comment by VictorVolpe** [comment]
> Transmission 4.0.6 has been blacklisted on at least one of the trackers I use regularly. After it was already fixed as well. I don't suppose we could get a 4.0.7 release to get past that version issue.

Where was it fixed? I didn't see any commit regarding this issue. I'm running the latest version of transmission-daemon 4.1.0-dev (fea9a8efb7) and still getting SIGPIPE. This is messed up! A complete nightmare.

**Comment by writequit** [comment]
It was fixed in curl, and that has found its way to debian unstable already.

**Comment by VictorVolpe** [comment]
> It was fixed in curl, and that has found its way to debian unstable already.

I use FreeBSD, but thanks for the information.

**Comment by xor-gate** [comment]
> > It was fixed in curl, and that has found its way to debian unstable already.
> 
> I use FreeBSD, but thanks for the information.

I'm using FreeBSD also. If you have a clean install you are bit out of luck probably (as I have not found out to cherry-pick .pkg files from a mirror). I updated my system (`freebsd-update`) and packages (`pkg update && pkg upgrade`). Using `13.3-RELEASE-p5` and `transmission-daemon 4.0.6 (38c164933e)`. I removed curl with `pkg` and installed from the cache with `pkg add /var/cache/pkg/curl-8.9.0~f08bc93756.pkg`. The corresponding [FreeBSD ticket #280644 is here](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=280644).

**Comment by imaami** [comment]
> It was fixed in curl, and that has found its way to debian unstable already.

There is a commit in the curl repo that fixes it, but there's no new curl release yet. Technically it's not *fixed in curl* until it's actually part of a release, although many distros have already added that one commit as a fix patch.

**Comment by xor-gate** [comment]
> > It was fixed in curl, and that has found its way to debian unstable already.
> 
> There is a commit in the curl repo that fixes it, but there's no new curl release yet. Technically it's not _fixed in curl_ until it's actually part of a release, although many distros have already added that one commit as a fix patch.

In FreeBSD ports there is no local patch for curl (currently). See https://cgit.freebsd.org/ports/log/ftp/curl. So we have to be a little more patient.

**Comment by edgeman16** [comment]
> > > It was fixed in curl, and that has found its way to debian unstable already.
> > 
> > 
> > There is a commit in the curl repo that fixes it, but there's no new curl release yet. Technically it's not _fixed in curl_ until it's actually part of a release, although many distros have already added that one commit as a fix patch.
> 
> In FreeBSD ports distro there is no local patch for curl (currently). See https://cgit.freebsd.org/ports/log/ftp/curl. So we have to be a little more patient.

FreeBSD ports tree has patched curl as of this commit
https://cgit.freebsd.org/ports/commit/?id=73d53fec20e010a33757af558db532d30c3a085b

**Comment by xor-gate** [comment]
> > In FreeBSD ports distro there is no local patch for curl (currently). See https://cgit.freebsd.org/ports/log/ftp/curl. So we have to be a little more patient.
> 
> FreeBSD ports tree has patched curl as of this commit https://cgit.freebsd.org/ports/commit/?id=73d53fec20e010a33757af558db532d30c3a085b

The corresponding ticket in the FreeBSD bugtracker has also been closed. https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=280644

**Comment by Pentaphon** [comment]
@Krandor1 please discuss the libcurl issue here. It is not a Transmission problem. It an issue with whoever packages your instance of Transmission and more trackers will lift the ban in a week or two.

**Comment by Krandor1** [comment]
Glad to hear that and hopefully the bans will be removed soon.  For now at
least had to downgrade but my tracker did indicate it was just temporary so
hopefully it won't last long.
--
Brian Hartsfield

**Comment by Tharn** [comment]
Don't count on 4.0.6 being whitelisted again, there is another big issue with it.

**Comment by KevinColemanInc** [comment]
Why aren't they bumping the version? Otherwise, how do the trackers know if a version was impacted?

**Comment by fetzu** [comment]
> Don't count on 4.0.6 being whitelisted again, there is another big issue with it.

Which "big issue" are you talking about?

**Comment by dechamps** [comment]
> > Don't count on 4.0.6 being whitelisted again, there is another big issue with it.
> 
> Which "big issue" are you talking about?

They're referring to #7078.

**Comment by garoto** [comment]
> Why aren't they bumping the version? Otherwise, how do the trackers know if a version was impacted?

The trackers don't know and can't know. The widespread bans of v4.0.6 is only because that is the most likely revision number to be affected by the libcurl issue, since some deployable solutions are building using master of the dependencies I guess? One can build a upcoming release (or a previous one) against the offending libcurl commit and that would as well have the buggy behavior described. Hope that helps.

**Comment by KevinColemanInc** [comment]
The transmission could be updated to prevent building with a bad version of libcurl. If it detects the bad version during the build, it fails or reports the libcurl version being used.

**Comment by tearfur** [comment]
@KevinColemanInc How do you suggest we detect a locally patched version of 8.9.1, like what Debian Unstable is now using?

From https://salsa.debian.org/debian/curl/-/compare/debian%2F8.9.1-1...debian%2F8.9.1-2?from_project_id=5916&straight=true, it seems to me that nor did the libcurl version detected by CMake or `LIBCURL_VERSION_NUM` change.

**Comment by tearfur** [comment]
We better wait until `8.10.0` is released before making that change.

**Comment by proton-ab** [comment]
> Given that there's nothing in the BitTorrent spec that says how long trackers should wait before deleting a peer, I'm not too sure how re-using the same peer_id with its associated stats after a long pause in announcing should be handled.

Officially BEP-3 is mute on the issue and just says that clients should generate new one per download - whether they store that later for resumption is hence assumed to be left to implementation. While Transmission can of course say that it is trackers that are technically expecting more than specs mandate, it would be an uphill battle if they decide to ban Transmission for this reason. I think we should steer on the caution side here and not change well established behavior that other clients exhibit.

**Comment by imaami** [comment]
> Transmission could be updated to prevent building with a bad version of libcurl. If it detects the bad version during the build, it fails or reports the libcurl version being used in the User-Agent, enabling trackers to continue to block bad builds.

Building transmission against libcurl 8.9.1 doesn't mean the build result is buggy. It would work just fine on a system with an unaffected libcurl. Practically all Linux distros link everything dynamically. Do a runtime [curl_version_info()](https://curl.se/libcurl/c/curl_version_info.html) call instead of baking it in at compile time.

**Comment by dechamps** [comment]
> Do a runtime [curl_version_info()](https://curl.se/libcurl/c/curl_version_info.html) call instead of baking it in at compile time.

That may result in false positives due to some distros sadly [fixing libcurl without bumping the version number](https://sources.debian.org/patches/curl/8.9.1-2/ignore-SIGPIPE-after-init.patch/). That said, false positives, being merely an user inconvenience, are arguably less of a concern than false negatives that lead to DDoS'ing trackers.

**Comment by vincejv** [comment]
@dechamps then perhaps do a runtime check if the bug exists rather than just a version string check

Runtime check meaning try to reproduce the bug through unit tests or some sort during compilation

**Comment by imaami** [comment]
> @dechamps then perhaps do a runtime check if the bug exists rather than just a version string check
> 
> Runtime check meaning try to reproduce the bug through unit tests or some sort during compilation

A runtime check is the opposite of a check during compilation.

The libcurl installed on the build system at compilation time has no bearing on what any given user has installed. And the libcurl binary can (and will) change over time. The base assumption should be that every time you start transmission, the libcurl version is unknown and has to be figured out.

Like already pointed out, distro-patched libcurl builds cause false positives because the runtime version info says 8.9.1 just like the unpatched one. The only way to reliably tell apart a patched 8.9.1 from a buggy one at runtime would involve something like this:

- Write a minimal program that reliably crashes without delay if it's run with a buggy libcurl 8.9.1, and exits with success if run with a patched libcurl 8.9.1 (or an unproblematic different version). This of course requires some development effort.
- When transmission starts, check `curl_version_info()`. If libcurl is not 8.9.1, everything is OK and no further tests are needed.
- If libcurl is 8.9.1, fork a child process that tries to provoke the crash. Call `waitpid()` in the parent.
- If libcurl is patched `waitpid()` should see a normal exit, if libcurl is buggy the parent should see exit code 141.
- If the child process exited abnormally, modify transmission's user agent string e.g. by changing libcurl version to 8.9.0 or 8.9.2. (The former is probably a better choice because 8.9.2 would be entirely fictional.)

I don't know if the SIGPIPE can be triggered on demand (immediately) or if it's probabilistic. If it's the latter then that makes implementing a runtime test somewhat less fun. It would probably not be smart to do nothing for 30 seconds during start.

**Comment by ivanjx** [comment]
@imaami in my case the crash happened after 15-30 minutes. your idea would not work.

**Comment by vincejv** [comment]
@imaami
> The libcurl installed on the build system at compilation time has no bearing on what any given user has installed.

This should be changed, are you referring to libcurl being dynamically linked instead of statically linked and bundled into the executable?

> libcurl binary can (and will) change over time. The base assumption should be that every time you start transmission, the libcurl version is unknown and has to be figured out.

Then Just statically link libcurl so there are more eyes on board when changes come to libcurl making sure transmission is compatible with libcurl changes and patches and work around the bugs it might introduce on how transmission invokes the library

**Comment by vincejv** [comment]
@imaami please correct me if I'm wrong but here's what I understand from your reply

> The libcurl installed on the build system at compilation time has no bearing on what any given user has installed. And the libcurl binary can (and will) change over time. The base assumption should be that every time you start transmission, the libcurl version is unknown and has to be figured out.

are you suggesting that libcurl is currently a dynamically linked library and not statically linked i.e. the library is not embedded in the final executable? suppose that's the case I **strongly believe** we should change libcurl to be statically linked and embedded to transmission. In that case, that is if you can't figure out how to do a runtime check, which based on your and other people's replies, is infeasible as reproducing the issue would require the client to do nothing for several minutes.

this solves many issues going forward like ABI compatibility, and other future changes in libcurl that will break transmission. It makes it so that libcurl version is tightly bound and dependent on the transmission version, and more eyes (reviewers, testers) to check if the current libcurl code base is compatible with transmission before upgrading the libcurl library.

The only downside is transmission would grow bigger by a few megabytes.

**Comment by dechamps** [comment]
> The only downside is transmission would grow bigger by a few megabytes.

That's not the only downside. The other downside is updates to libcurl would be tied to Transmission releases. Say a security vulnerability is found in libcurl - something that could definitely happen given it is very much a network-facing library. Distros will quickly push updates to libcurl, but Transmission may lag behind and stay vulnerable for longer. Same for bug fixes (the flip side of regressions like this one).

In any case, I think Linux distributions would have something to say about libraries being packaged statically vs. dynamically (e.g. in Debian, everything is dynamically linked AFAIK). It's probably them you'd need to convince. Their first reaction would probably be to ask why Transmission is such a special case (this could have happened to literally any software using libcurl), and if you tell them it's not, then they're going to point out that by that logic everything should link everything statically, and let you ponder the consequences in terms of resource usage and security fixes. Expect an uphill battle.

**Comment by tearfur** [comment]
> what do ardvin's libtorrent and rTorrent use?

FYI, libtorrent uses a bespoke async HTTP client based on `Boost.Asio` (`src/http_tracker_connection.cpp` for anyone interested). We currently rely on `std::thread` and curl's multi interface to achieve a similar effect (`libtransmission/web.cc` for anyone interested).

No idea about rtorrent.

**Comment by proton-ab** [comment]
> and only transmission of all clients got banned

Bans on Transmission were fueled entirely by pure ignorance of team behind `linuxserver/docker-transmission` docker images which are updated automatically and track `alpine:edge` as source. They do not even go to effort to build Transmission from source - they install it from Alpine package manager!

I suggest we close this issue and just do nothing, because there is nothing we can do. Next time Transmission gets banned it could be because above mentioned docker image pulled broken version of `musl` from experimental Alpine repos.

**Comment by vincejv** [comment]
@proton-ab 
They use bleeding edge versions of dependencies including musl when building qBittorrent/libtorrent too and it didn't get banned at all, and I don't see it being an issue there despite being the #1 client across several trackers if not ALL trackers

It's already clarified on https://github.com/linuxserver/docker-transmission/issues/286 that they won't be moving to stable as quoted "_there won't be transmission updates for 6 months at a time, which was the reason it was moved to edge in the first place._" so **why is transmission exempted** when qBittorrent is built from bleeding edge but no complaints over there being the number 1 torrent client based on usage.

> nothing we can do

I'd still like to be convinced why boost.io is not an option going forward and remain with libcurl, i understand it's a huge development effort but it's best be justified and no rush anyway in migrating away from libcurl.

**Comment by tearfur** [comment]
@vincejv IMO, it's not that Boost is absolutely not an option, but we need to have a ***very*** good reason to do it. As you probably understand, it's a huge effort to make the switch, we'd have to rebuild a lot of our infrastructure.

Bugs like that can happen to any project, and what we should do going foward is to stick with a solution that's reasonably trustworthy and stable. Do you have any reason to believe curl is much more vulnerable than Boost (or other alternatives in general), such that switching libraries is justified? I don't see any myself.

* * *

On a separate note, based on everyone's discussion, it seems like there's no "definitive" mitigation we can do about this issue. Maybe the most we can do is print a warning in the logs if the curl version number is `8.9.1`.

**Comment by ivanjx** [comment]
maybe transmission should just keep track of last few start times and check if they are frequent enough. if so then just dont connect to anything and display error message somehow.

**Comment by imaami** [comment]
> @imaami in my case the crash happened after 15-30 minutes. your idea would not work.

It might not work, that's true. I haven't had a close look at the relevant code in Transmission or experimented with triggering the libcurl bug, so what I can say for sure is based on how shared libraries behave.

It's not possible to predict or guarantee the exact identity of the runtime libcurl.so when building a program that links against it. This is a fundamental property of dynamic linking. And although you *can* get the actual libcurl version being used during runtime by calling `curl_version_info()`, other people in this discussion have said that the returned version info of a patched 8.9.1 is identical to an unpatched one.

With all of the above, there's really only one reliable way left: a runtime "black box" experiment. Whether or not it's possible depends on specifics I don't have time to research. (I can type out these comments here every now and then but that's about it, I'm not a Transmission or libcurl developer.)

*<end of generally true stuff, the rest is guessing and speculation>*

It's unsurprising that there are individual differences in how long it takes for the crash to occur. The problem is in the connection lifetime handling code, so the number and frequency of p2p connections affects it a lot. And—this is just speculation on my part—it's not uncommon to see behavior like this being caused by a race condition during signal handling.

**Comment by ivanjx** [comment]
imo the solutions provided here are mostly overcomplicated or requires the others like the lsio docker to change how they work. believe me that none would fix the issue.

instead we should just detect whether the transmission is being restarted frequent enough and just trigger “safe mode”.

**Comment by proton-ab** [comment]
> They use bleeding edge versions of dependencies including musl when building qBittorrent/libtorrent too and it didn't get banned at all, and I don't see it being an issue there despite being the https://github.com/transmission/transmission/pull/1 client across several trackers if not ALL trackers

It just so happens no bug affected them... yet :> Believe me though that plenty of trackers would do it should some crash affect them too; it's not like they aren't annoyed already at the broken BEP-7 implementation in arvidn's libtorrent causing various edge cases weirdness thanks to its fruitless attempts to bind to specific IPs for sending announce requests.

**Comment by MoonSweep** [comment]
If I may...

As an experienced sysadmin, I think that this whole discussion is getting out of hand, and more importantly, for nothing. The problem is not in Transmission (bugs happen, and they get fixed; that's what `libcurl` did), but in the fact that **the service was restarted repeatedly as soon as it crashed**.

I'm an experienced Debian sysadmin, so I neither use `Alpine`, nor `OpenRC`, nor LinuxServer.io's `docker-transmission`, but in `systemd`, a service is restarted only in very specific conditions. By default, it's restarted at most 5 times *if it was unable to start*. If it crashes *during run*, it's not restarted at all, unless **specific configuration was added by the user**. By experience (I admit I didn't check), it would be very surprising that `Alpine`'s or `OpenRC`'s default behavior was to endlessly restart a service which crashes during run. Maybe it's `docker-transmission`, which does, or `Docker` itself, or the users of this pre-built container, I don't know, and honestly, I don't care (I never use pre-built containers, I make my own images when I need one).

When a program is crashing, this means there is an underlying bug which makes it crash, and the right thing to do is to find this bug and fix it, not dumbly restart the service repeatedly. It's clearly not Transmission's responsibility to do anything about that fact, and certainly not waste precious computing time to log its own start times and change its behavior if it thinks it restarts too often. The culprit brick is way, way down the chain, whether it's `OpenRC`, or `Alpine`, or `LinuxServer.io`, or `Docker` (again, I didn't search), it's not Transmission's fault if the service was infinitely restarted until trackers decided to ban it.

Not mentioning the fact that the problem is already fixed in `libcurl`, they'll release a fixed version soon, most major distributions cared to patch the previous version in the meantime, and in a couple of weeks/months, this bug will be long forgotten, leaving those checks, which only gets triggered in one out of a million runs, rotting in Transmission's source code, until in 10 years developers will ask themselves "why the hell did we implement those useless checks?" and get rid of them.

So, if I may suggest a piece of advice, this discussion should be closed, and you should put your efforts into maintaining and enhancing Transmission, instead of wasting your time trying to cover for dumb admins' errors.

**Comment by tearfur** [comment]
@MoonSweep I've been meaning to close this issue as well, but I felt the need to make sure people understand the reasons for us to not do anything about it.

Thanks for your write up. Hopefully that will clear things up more.

**Comment by KevinColemanInc** [comment]
@MoonSweep I understand what you're saying, but I think you have misunderstood the problem. 

The problem is not that transmission crashes or is disruptive to trackers. The problem is that trackers have banned Transmission's User-Agent, making the product completely useless.

You can point fingers all you want at other people claiming this is their problem and not Transmissions. But at the end of the day, I can't use Transmission to download or seed torrents right now.

**Comment by Tharn** [comment]
The problem is that there's a constellation of other software that makes Transmission crash and then restart in a loop. It's up to docker and libcurl respectively not to have either awful restart-on-crash defaults, or cause programs to crash. The problem is still expressed with Transmission... but similarly, people who use it on Windows or people who use it with fixed version libaries and not the problematic curl version will never run into that problem. 4.0.6 is still banned for them. What can Transmission do? Simply push a new version as fast as possible.

**Comment by MoonSweep** [comment]
@KevinColemanInc @Tharn 

Your complaints finally got me interested enough to check the chain, and the culprit is obviously [linuxsever/docker-transmission](https://github.com/linuxserver/docker-transmission). Not only are they intent to use `Alpine` edge instead of stable (as already stated several times in this thread), but the `docker-compose` file also has `restart: unless-stopped` (`Docker`'s default is `no`). 

Now I'm entering rant mode.

As a sysadmin, I qualify this combo as a pure heresy. This is asking for trouble. The people behind this tickling time bomb clearly don't know what they're doing (probably developers with very poor admin skills), and this is also valid for people using such bad software in a production environment (and I'm not talking about companies here; if you can't afford downtimes, this is considered production, even for individual systems like your own seedbox).

**You don't use the development branch in production, period.**

There's a reason why every major distribution has development and stable branches, it's precisely for a very small number of tech-savvy users to use the development branch, to help find and fix bugs, so that those bugs don't reach the stable branch, which can be safely used by the majority of people. That's what I do, I use Debian unstable, but only on my home desktop PC. Every other system (laptop, servers, etc) has the stable version, because I can't afford them to suddenly stop working or even exhibit the slightest problem after an update. That's how people found why `Transmission` suddenly started to crash (seemingly) randomly, and that's why I reported the bug against `libcurl` to Debian, and that's how Debian unstable was fixed, so that Debian stable won't suffer from this bug when it's released.

 If `linuxsever/docker-transmission` were not so lazy, they would use `Alpine` stable branch, and compile `Transmission`'s latest version under it (which sometimes demands some effort, but not always). That's called *backporting* and it's a very common practice for Debian and its derivatives (and probably other distros I guess). If they did that, their container would have run the latest `Transmission`, but compiled against a stable `libcurl` version, and the whole mess would never have happened. But no, they had to use **the whole `Alpine edge` branch** just to have the latest `Transmission`, and on top of that, set the container to restart automatically whatever the reason it failed.

Now that I know more, I think you're right, Transmission should indeed be modified, but to check if it's running under `linuxsever/docker-transmission` (or maybe even anything remotely related to `linuxserver`, I don't know if the same team publish all their containers) and plainly refuse to start. Then you could nicely ask trackers to whitelist `Transmission` 4.0.7+ and everybody would be happy.

End of rant.

**Comment by vincejv** [comment]
> but compiled against a stable libcurl version, and the whole mess would never have happened

sorry, are you saying the buggy libcurl was not released on the stable tree? Sure you are correct edge is a development version but I'm not so sure if libcurl used was from the development branch, afaik the 8.9.1 was released as "stable", but glad to be corrected if i'm wrong

**Comment by MoonSweep** [comment]
@vincejv

You need to learn how distributions work. Let me explain a bit.

In distributions' vernacular, `development`, `unstable`, `edge` or whatever doesn't mean that a this branch uses development branches of every piece of software it contains. It means that they use the *latest* (most of the time, stable) versions of those pieces, and when all those pieces are tested to work well *together*, the whole release is deemed as `stable`, and published.

In this case, Alpine Linux does use the stable version of `libcurl` in its `edge` branch. According to their [Git repository](https://git.alpinelinux.org/aports/log/?qt=grep&q=curl), they pushed 8.9.1 in edge the day it was released (July 31) and patched it on August 8 to fix the bug.

As for their stable branch, still according to their Git repository, v3.20.2 (released July 22) still had `libcurl` [v8.8.0](https://git.alpinelinux.org/aports/log/?qt=grep&q=curl&h=v3.20.2), whereas v3.20.3 (released September 6, today) has [a patched version of 8.9.1](https://git.alpinelinux.org/aports/log/?qt=grep&q=curl&h=v3.20.3).

So, yeah, had `linuxserver.io` used the stable branch of Alpine to build their container image, this mess would never have happened.

**Comment by vincejv** [comment]
> So, yeah, had linuxserver.io used the stable branch of Alpine to build their container image, this mess would never have happened.

@MoonSweep 
Sorry it still doesn't make sense to me, let's say they change the base environment to **alpine:stable**, pull from the git sources of transmission and libcurl

pull the source tarballs:
transmission: https://github.com/transmission/transmission/releases/tag/4.0.6
curl: https://github.com/curl/curl/releases/tag/curl-8_9_1

and other dependencies, compile them, Transmission 4.0.6 and libcurl 8.9.1 on alpine stable, this wouldn't be an issue??

**Comment by MoonSweep** [comment]
@vincejv 

Why pull `curl` 8.9.1 ?

If the latest `transmission` builds successfully against the version of `libcurl` in the target release, there's no need to also pull the latest `libcurl`.

That's the whole point of *backporting* a package : use the libraries from the target release, so that your package integrates as seamlessly as possible in it.

Sometimes you also need to backport a newer version of some libraries, but I'd be very surprised if the versions in Alpine stable didn't satisfy the requirements of the latest `transmission`, and allow to build it.

**Comment by vincejv** [comment]
> Why pull curl 8.9.1 ?
If the latest transmission builds successfully against the version of libcurl in the target release, there's no need to also pull the latest libcurl.

it's marked as a stable release by the curl development team, and it may contain the latest security fixes. Unless explicitly specified by the dependency that its ABI has changed, it makes more sense to me to use latest instead of the backport.

Backports are not necessarily always the best, sometimes changes are missed during backports and even takes more time to backport it due to the complexity, you can take a look at Linux LTS kernel releases, several bugs are encountered from time to time due to incomplete backports.

**Comment by MoonSweep** [comment]
I won't waste more time with this.

**You don't use development branches in production, period.**

You make bad choices, and Transmission took a stray bullet because you f***ed up. Your containers are dangerous. Stop making excuses. Admit it and fix it, or stop pretending to do a job you obviously don't know anything about.

**Comment by vincejv** [comment]
> You don't use development branches in production, period.

I already explained the scenario wherein there were no development branches involved, only latest stable releases....

**Comment by vincejv** [comment]
curl 8.10.0 just dropped as stable, hope someone can do a sanity check if there's another severe issue that would cause to get transmission to get banned again

https://github.com/curl/curl/releases/tag/curl-8_10_0

**Comment by MoonSweep** [comment]
If you used the stable branch of your favorite distro, you wouldn't worry about that, and you wouldn't put Transmission in danger of being banned again.

**Comment by vincejv** [comment]
@MoonSweep what do you mean? Can't someone use the **latest stable branch of a distro** and compile against latest transmission stable and curl stable? Is it not advisable to compile from the **latest stable source** directly instead of using your distro's package manager?

**Comment by MoonSweep** [comment]
But why the hell do you absolutely want to backport curl alongside Transmission ? I explained you in detail why that's not necessary, and not wanted. Stop making excuses. Just build your container from Alpine stable, and build **only Transmission from edge** on top of it. That's all. I know that's a bit more work than getting the whole system from edge directly as binary, but that's the only way to do things properly, and that's how you avoid bugs and avoid "get Transmission banned". It's not Transmission's job to check if every new version of every library used by it introduces new flaws, that's the job of the distributions, and that's precisely why they all have development branches, which you're **not supposed to use on production systems**.

**Comment by vincejv** [comment]
There's a tracker that's reporting they are being DDOS'd, I hope it's not coming from this libcurl/transmission issue, or else I'd really push to these warez admins to do a blanket ban on future transmission versions unless the transmission maintainers are open to merge this PR at the very least: https://github.com/transmission/transmission/pull/7126 

Libcurl 8.9.xx+ is slowly rolling out to stable distros
Reference:
https://github.com/transmission/transmission/issues/7133

Quoting
> This is related to https://github.com/transmission/transmission/issues/7035. Someone at Alpine thought it is amazing idea to push 3.18 and 3.19 branches to 8.9.1 but without applicable patch that is present in 3.20 and main, so it is **wrecking havoc month after it was fixed.**

**Comment by MoonSweep** [comment]
What about ceasing to blame others for your mistakes and fix your container to use Alpine Stable and not restarting automatically after a crash ?

**Comment by vincejv** [comment]
@MoonSweep The buggy libcurl was shipped with Alpine Stable
if you read #7113 and the quoted message, it wasn't even repro inside a container, but in pop os.

> This is related to https://github.com/transmission/transmission/issues/7035. Someone at Alpine thought it is amazing idea to push 3.18 and 3.19 branches to 8.9.1 but without applicable patch that is present in 3.20 and main, so it is wrecking havoc month after it was fixed.

**Comment by vincejv** [comment]
@MoonSweep so idk why you are still rambling about containers and alpine stable

**Comment by MoonSweep** [comment]
The buggy libcurl was shipped with *old* 3.18/3.19 stable versions (because - I didn't check, but I guess - in those they just commit to fix security issues - if it's actually the reason , I'll concede it's a terrible practice). Latest 3.20 stable had the patched version, and 8.10.0/8.10.1 after that.

As for the problem in other distros, it's not caused by curl but by specific settings in the systemd service file which prevents the app to notify systemd that it started, due to apparmor restrictions (https://github.com/transmission/transmission/issues/6991). I suggest you re-read https://github.com/transmission/transmission/issues/7133 carefully and actually try to understand it.

Nevertheless, in those distros, the service is not set to automatically restart indefinitely, so it's most probably not the cause for the current DDoS.

Advertising the libcurl version in the User-Agent wouldn't solve anything, quite the contrary, it would be highly counter-productive by risking banning users running a perfectly working Transmission.

Now stop necroposting this issue and pushing this PR every time there's a problem with Transmission.

**Comment by dechamps** [comment]
> The buggy libcurl was shipped with _old_ 3.18/3.19 stable versions (because - I didn't check, but I guess - in those they just commit to fix security issues - if it's actually the reason , I'll concede it's a terrible practice).

This was an oversight on Alpine's part. [It's fixed now](https://github.com/transmission/transmission/issues/7133#issuecomment-2391185151), with my thanks to @ptrcnull for pushing the fix literally 15 minutes after I made them aware of it.

**Comment by MoonSweep** [comment]
> This was an oversight on Alpine's part.

Much more likely indeed.
